src/map.cpp

```diff
@@ -6910,6 +6910,112 @@ void map::restock_fruits( const tripoint &p, int time_since_last_actualize )
+
+    // How many turns to produce 1 charge (250 ml) of sap?
+    const int turns_season = DAYS( calendar::season_length() );
+    const int producing_length = int( 0.75f * turns_season );
+
+    const int turns_to_produce = producing_length / ( maple_sap_per_season * 4 );
+
+    // How long of this time_since_last_actualize have we been in the producing period (late winter, early spring)?
+    int time_producing = 0;
+
+    if( time_since_last_actualize >= calendar::year_turns() ) {
+        time_producing = producing_length;
+    } else {
+        // We are only procuding sap on the intersection with the sap producing season.
+        int early_spring_end = int( 0.5f * turns_season );
+        int late_winter_start = int( 3.75f * turns_season );
+
+        calendar last_actualize = calendar::turn - time_since_last_actualize;
+        int last_actualize_tof = last_actualize.turn_of_year();
+        bool last_producing = (
+            last_actualize_tof >= late_winter_start ||
+            last_actualize_tof < early_spring_end
+        );
+        int current_tof = calendar::turn.turn_of_year();
+        bool current_producing = (
+            current_tof >= late_winter_start ||
+            current_tof < early_spring_end
+        );
+
+        int non_producing_length = int( 3.25f * turns_season );
+
+        if( last_producing && current_producing ) {
+            if( time_since_last_actualize < non_producing_length ) {
+                time_producing = time_since_last_actualize;
+            } else {
+                time_producing = time_since_last_actualize - non_producing_length;
+            }
+        } else if ( !last_producing && !current_producing ) {
+            if( time_since_last_actualize > non_producing_length ) {
+                time_producing = time_since_last_actualize - non_producing_length;
+            }
+        } else if ( last_producing && !current_producing ) {
+            // We hit the end of early spring
+            if( last_actualize_tof < early_spring_end ) {
+                time_producing = early_spring_end - last_actualize_tof;
+            } else {
+                time_producing = calendar::year_turns() - last_actualize_tof + early_spring_end;
+            }
+        } else if ( !last_producing && current_producing ) {
+            // We hit the start of late winter
+            if( current_tof >= late_winter_start ) {
+                time_producing = current_tof - late_winter_start;
+            } else {
+                time_producing = int( 0.25f * turns_season ) + current_tof;
+            }
+        }
+    }
+
+    long new_charges = roll_remainder( (double)time_producing / turns_to_produce );
+    // Not enough time to produce 1 charge of sap
+    if( new_charges <= 0 ) {
+        return;
+    }
+
+    // Is there a proper container?
+    auto items = g->m.i_at( p );
+    for( auto &it : items ) {
+        if( ( it.is_bucket() || it.is_watertight_container() ) && (
```

Comment:
This check is identical to the one in iexamine.

I'd extract it to an item method. Something like `can_hold_liquid( const itype &type )`, which would look like:
```
bool bucket_can_hold( const itype &liquid_type ) {
return ( it.is_bucket() || it.is_watertight_container() ) && (
    it.is_container_empty() ||
    !it.is_container_full( true ) &&
    it.contents.front().type->id == liquid_type;
}
```

Or you could even have it return the remaining charges of the liquid it can hold, with 0 indicating it can't hold any for any reason (such as not being a bucket).

Though doesn't `it.get_remaining_capacity_for_liquid( sap, true );` return the correct value anyway? I mean, I think you could just replace both the check here and in the iexamine with:
```
long cap = it.get_remaining_capacity_for_liquid( sap, true );
if( cap > 0 )
```

Or would something stop working then?